### PR TITLE
add new arm template for role assignment example

### PR DESCRIPTION
**Pull request by zhaomuzhi**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.

**Comment by zhaomuzhi** [comment]
@bmoore-msft , the build failed with this error
Status Message: The client '5991c0de-1362-4164-8ac0-0ab3b91175ee' with object id '5991c0de-1362-4164-8ac0-0ab3b91175ee' does not have authorization to perform action 'Microsoft.ManagedIdentity/userAssignedIdentities/write' over scope '/subscriptions/0cec7090-2e08-4498-9337-eb96ade50821/resourcegroups/azdo-84f20e67-fcd0-4147-827c-57614ce028d4-prereqs/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ci1927d0daf4da4e53' or the scope is invalid. If access was recently granted, please refresh your credentials. (Code:AuthorizationFailed)

https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=40969&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a

Do you know how I can overcome it?

**Comment by Blackmist** [comment]
Thanks @bmoore-msft , it's on our radar. Once it's successfully merged we will add/update reference from our Azure RBAC docs.

**Comment by bmoore-msft** [comment]
@zhaomuzhi - it's pretty hard to follow this, I think the first change is to use resource names and not full resourceIds for the parameters... that will make it easier to see what you're trying to do (and easier for users to supply values for).  The functions likely become extraneous and can be removed.  Let's try debugging with that change first...

**Comment by zhaomuzhi** [comment]
> @zhaomuzhi - it's pretty hard to follow this, I think the first change is to use resource names and not full resourceIds for the parameters... that will make it easier to see what you're trying to do (and easier for users to supply values for). The functions likely become extraneous and can be removed. Let's try debugging with that change first...

For your suggestion of using resource name instead of full resourceIds, my motivation of using resourceIds instead is because the dependent resource, e.g. storage account, may exist in different resource group or even different subscription. So if we don't ask customer to provide resourceId, they need to give Sub/RG/ResourceName for each dependent resource, and we have 6 such  dependent resources. So I feel it would be more user friendly for customer to pass resourceId instead of Sub/RG/ResourceName individually, and I wrote some functions to parse Sub/RG/ResourceName out of the resourceId.

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.

**Comment by bmoore-msft** [comment]
Supplying sub/rg/name is the common pattern, no human can reliably enter a resourceId...  As well you can default things like sub (and RG) since cross sub resource use is rare...  So we still need that change.

Additionally, if you really want to simplify, you can provide a createUIdefinition file and use the picker controls (or role your own) for resource selection...

**Comment by zhaomuzhi** [comment]
> Supplying sub/rg/name is the common pattern, no human can reliably enter a resourceId... As well you can default things like sub (and RG) since cross sub resource use is rare... So we still need that change.
> 
> Additionally, if you really want to simplify, you can provide a createUIdefinition file and use the picker controls (or role your own) for resource selection...

passed rg/name instead

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+                },
+                "condition": "[parameters('assignWorkspaceStorageAccount')]"
+              },
+              {
+                "type": "Microsoft.Authorization/roleAssignments",
+                "apiVersion": "2020-04-01-preview",
+                "name": "[variables('workspaceStorageAccountBlobDataContributorRoleName')]",
+                "scope": "[concat('Microsoft.Storage/storageAccounts/', parameters('workspaceStorageAccountName'))]",
+                "properties": {
+                  "roleDefinitionId": "[concat(variables('workspaceStorageAccount'), '/providers/Microsoft.Authorization/roleDefinitions/', variables('storageBlobDataContributorRoleDefinition'))]",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:242]: remove concat and use resourceId()

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+      },
+      {
+        "apiVersion": "2020-06-01",
+        "name": "databrickTemplate",
+        "type": "Microsoft.Resources/deployments",
+        "resourceGroup": "[parameters('workspaceDataBrickResourceGroup')]",
+        "properties": {
+          "mode": "Incremental",
+          "template": {
+            "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:260]: update schema

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            "parameters": {},
+            "variables": {},
+            "resources": [
+              {
+                "type": "Microsoft.Authorization/roleAssignments",
+                "apiVersion": "2020-04-01-preview",
+                "name": "[variables('workspaceDataBrickContributorRoleName')]",
+                "scope": "[concat('Microsoft.Databricks/workspaces/', parameters('workspaceDataBrickName'))]",
+                "properties": {
+                  "roleDefinitionId": "[concat(variables('workspaceDataBrick'), '/providers/Microsoft.Authorization/roleDefinitions/', variables('contributorRoleDefinition'))]",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:271]: remove concat() and use resourceId()

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+            "parameters": {},
+            "variables": {},
+            "resources": [
+              {
+                "type": "Microsoft.Authorization/roleAssignments",
+                "apiVersion": "2020-04-01-preview",
+                "name": "[variables('workspaceContainerRegistryContributorRoleName')]",
+                "scope": "[concat('Microsoft.ContainerRegistry/registries/', parameters('workspaceContainerRegistryName'))]",
+                "properties": {
+                  "roleDefinitionId": "[concat(variables('workspaceContainerRegistry'), '/providers/Microsoft.Authorization/roleDefinitions/', variables('contributorRoleDefinition'))]",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:300]: remove concat() and use resourceId()

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+      },
+      {
+        "apiVersion": "2020-06-01",
+        "name": "containerRegistryTemplate",
+        "type": "Microsoft.Resources/deployments",
+        "resourceGroup": "[parameters('workspaceContainerRegistryResourceGroup')]",
+        "properties": {
+          "mode": "Incremental",
+          "template": {
+            "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:289]: update schema

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+      },
+      {
+        "apiVersion": "2020-06-01",
+        "name": "cmkKeyvaultTemplate",
+        "type": "Microsoft.Resources/deployments",
+        "resourceGroup": "[parameters('workspaceCMKKeyVaultResourceGroup')]",
+        "properties": {
+          "mode": "Incremental",
+          "template": {
+            "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:318]: update schema

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+            "parameters": {},
+            "variables": {},
+            "resources": [
+              {
+                "type": "Microsoft.Authorization/roleAssignments",
+                "apiVersion": "2020-04-01-preview",
+                "name": "[variables('workspaceCMKKeyVaultContributorRoleName')]",
+                "scope": "[concat('Microsoft.KeyVault/vaults/', parameters('workspaceCMKKeyVaultName'))]",
+                "properties": {
+                  "roleDefinitionId": "[concat(variables('workspaceCMKKeyVault'), '/providers/Microsoft.Authorization/roleDefinitions/', variables('contributorRoleDefinition'))]",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:329]: remove concat() and use resourceId()

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+      },
+      {
+        "apiVersion": "2020-06-01",
+        "name": "workspaceResourceGroupTemplate",
+        "type": "Microsoft.Resources/deployments",
+        "resourceGroup": "[parameters('workspaceResourceGroup')]",
+        "properties": {
+          "mode": "Incremental",
+          "template": {
+            "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.json:347]: update schema

File: 201-machine-learning-dependencies-role-assignment/metadata.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    "$schema": "https://aka.ms/azure-quickstart-templates-metadata-schema#",
+    "type": "QuickStart",
+    "itemDisplayName": "User assigned identity role assignment template",
+    "description": "A template that creates role assignments of user assigned identity on resources that Azure Machine Learning workspace depends on",
+    "summary": "Demonstrates What role assignment are needed on the dependent resources of Azure Machine Learning user assigned identity enabled workspace to make the workspace functions as expected.",
+    "githubUsername": "zhaomuzhi",
+    "docOwner": "Blackmist",
+    "dateUpdated": "2020-12-03",
+    "environments": [
+        "AzureCloud"
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/metadata.json:11]: why not USGov?

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.parameters.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (bmoore-msft) [inline 201-machine-learning-dependencies-role-assignment/azuredeploy.parameters.json:1]: let's add parameter values here that will test all the resources in the template (i.e. change the false to true on the flags)...  you'll need to add more to the prereq template as well...

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+      "workspaceStorageAccountContributorRoleName": "[guid(variables('workspaceStorageAccount'), 'Contributor')]",
+      "workspaceStorageAccountBlobDataContributorRoleName": "[guid(variables('workspaceStorageAccount'), 'StorageBlobDataContributor')]",
+      "workspaceKeyVaultAdministratorRoleName": "[guid(variables('workspaceKeyVault'), 'KeyVaultAdministrator')]",
+      "workspaceContainerRegistryContributorRoleName": "[guid(variables('workspaceContainerRegistry'), 'Contributor')]",
+      "workspaceDataBrickContributorRoleName": "[guid(variables('workspaceDataBrick'), 'Contributor')]",
+      "workspaceCMKKeyVaultContributorRoleName": "[guid(variables('workspaceCMKKeyVault'), 'Contributor')]"
+    },
+    "resources": [
+      {
+        "apiVersion": "2020-06-01",
```
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:158]: fixed

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+              {
+                "type": "Microsoft.KeyVault/vaults/accessPolicies",
+                "apiVersion": "2019-09-01",
+                "name": "[concat(parameters('workspaceKeyVaultName'), '/add')]",
+                "properties": {
+                  "accessPolicies": [
+                    {
+                      "tenantId": "[reference(variables('primaryUserAssignedIdentity'), '2018-11-30').tenantId]",
+                      "objectId": "[reference(variables('primaryUserAssignedIdentity'), '2018-11-30').principalId]",
+                      "applicationId": "[reference(variables('primaryUserAssignedIdentity'), '2018-11-30').clientId]",
```
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:179]: which one is redundant?

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+    "resources": [
+      {
+        "apiVersion": "2020-06-01",
+        "name": "keyvaultTemplate",
+        "type": "Microsoft.Resources/deployments",
+        "resourceGroup": "[parameters('workspaceKeyVaultResourceGroup')]",
+        "properties": {
+          "mode": "Incremental",
+          "template": {
+            "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
```
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:165]: updated to 2019-04-01

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+            "parameters": {},
+            "variables": {},
+            "resources": [
+              {
+                "type": "Microsoft.Authorization/roleAssignments",
+                "apiVersion": "2020-04-01-preview",
+                "name": "[variables('workspaceStorageAccountContributorRoleName')]",
+                "scope": "[concat('Microsoft.Storage/storageAccounts/', parameters('workspaceStorageAccountName'))]",
+                "properties": {
+                  "roleDefinitionId": "[concat(variables('workspaceStorageAccount'), '/providers/Microsoft.Authorization/roleDefinitions/', variables('contributorRoleDefinition'))]",
```
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:231]: switched to use extensionResourceId

File: 201-machine-learning-dependencies-role-assignment/metadata.json (hunk 9 above)
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/metadata.json:11]: The UAI feature has not been enabled for sovereign cloud yet. We will enable it on service side later.

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+        "defaultValue": "",
+        "type": "String",
+        "metadata": {
+          "description": "Resource name of keyvault that hosts data encryption key of cmk workspace."
+        }
+      }
+    },
+    "variables": {
+      "primaryUserAssignedIdentity": "[resourceId(parameters('primaryUserAssignedIdentityResourceGroup'), 'microsoft.managedidentity/userassignedidentities', parameters('primaryUserAssignedIdentityName'))]",
+      "workspaceResourceGroup": "[concat(subscription().id, '/resourceGroups', parameters('workspaceResourceGroup'))]",
```
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json]: I tried [subscriptionResourceId('resourceGroups', parameters('workspaceResourceGroup'))], but it doesn't work, concat is the only way I know about how to get ARM Id of a resource group...

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.parameters.json (hunk 10 above)
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.parameters.json:1]: done

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+      "workspaceStorageAccount": "[resourceId(parameters('workspaceStorageAccountResourceGroup'), 'microsoft.storage/storageaccounts', parameters('workspaceStorageAccountName'))]",
+      "workspaceContainerRegistry": "[resourceId(parameters('workspaceContainerRegistryResourceGroup'), 'Microsoft.ContainerRegistry/registries', parameters('workspaceContainerRegistryName'))]",
+      "workspaceDataBrick": "[resourceId(parameters('workspaceDataBrickResourceGroup'), 'Microsoft.Databricks/workspaces', parameters('workspaceDataBrickName'))]",
+      "workspaceCMKKeyVault": "[resourceId(parameters('workspaceCMKKeyVaultResourceGroup'), 'microsoft.keyvault/vaults', parameters('workspaceCMKKeyVaultName'))]",
+      "contributorRoleDefinition": "b24988ac-6180-42a0-ab88-20f7382dd24c",
+      "storageBlobDataContributorRoleDefinition": "ba92f5b4-2d11-453d-a403-e96b0029c9fe",
+      "keyVaultAdministratorRoleDefinition": "00482a5a-887f-4fb3-b363-3b7fe8e74483",
+      "workspaceResourceGroupContributorRoleName": "[guid(variables('workspaceResourceGroup'), 'Contributor')]",
```
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:148]: done

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json]: subscriptionResourceId('Microsoft.Resources/resourceGroups', parameters('workspaceResourceGroup'))

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:178]: applicationId - you're using an MSI not an app registration.

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (bmoore-msft) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json]: just use:
 resourceId('Microsoft.Authorization/roleDefinitions', variables('contributorRoleDefinition'))

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json:178]: Okay

File: 201-machine-learning-dependencies-role-assignment/azuredeploy.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (zhaomuzhi) [reply 201-machine-learning-dependencies-role-assignment/azuredeploy.json]: extensionResourceId works well. Can I keep it?

**Comment by bmoore-msft** [comment]
@Blackmist - check this PR for updates that may be needed to documentation that references this sample.  [this is an automated message]
